lib/puppet/functions/tree_each.rb

```diff
@@ -0,0 +1,166 @@
+# Runs a [lambda](http://docs.puppetlabs.com/puppet/latest/reference/lang_lambdas.html)
+# recursively and repeatedly using each value in a data structure, then returns the values unchanged.
+#
+# This function takes two mandatory arguments and one optional, in this order:
+#
+# 1. An `Array`, `Hash`, or `Iterator` that the function will iterate over.
+# 2. An optional hash with the options:
+#    * include => `Optional[T]` # only yield values matching the type, default `Any` (include all)
+#    * exclude => `Optional[T]` # only yield values that do not match the filter type, default `undef` (exclude none)
+# 3. A lambda, which the function calls for each element in the first argument. It must
```

Comment:
I assume that `exclude` has higher priority than `include`? In any case, the priority should be stated.